File: speed/concurrency/concurrency.jl
using Base.Threads
using HTTP
using JSON
using Printf

# parallel http requests test using Julia's @async and @sync
function parallel_http_test(num_requests::Int)::Float64
    start_time = time_ns()
    
    successful = Threads.Atomic{Int}(0)
    
    @sync begin
        for i in 1:num_requests
            @async begin
                try
                    response = HTTP.get("http://127.0.0.1:8000/fast", connect_timeout=5, readtimeout=5)
                    if response.status == 200
                        Threads.atomic_add!(successful, 1)
                    end
                catch
                    # ignore errors for benchmark
                end
            end
        end
    end
    
    end_time = time_ns()
    _ = successful[]  # prevent optimization
    return (end_time - start_time) / 1_000_000
end

# producer-consumer queue test using Julia's channels
function producer_consumer_test(num_pairs::Int, items_per_thread::Int)::Float64
    start_time = time_ns()
    
    task_channel = Channel{Int}(1000)
    processed = Threads.Atomic{Int}(0)
    
    # create producer tasks
    producer_tasks = []
    for i in 1:num_pairs
        task = @async begin
            for j in 1:items_per_thread
                put!(task_channel, (i-1) * 1000 + j)
            end
        end
        push!(producer_tasks, task)
    end
    
    # create consumer tasks
    consumer_tasks = []
    for i in 1:num_pairs
        task = @async begin
            for j in 1:items_per_thread
                item = take!(task_channel)
                
                # simulate processing
                dummy = item * item
                
                Threads.atomic_add!(processed, 1)
            end
        end
        push!(consumer_tasks, task)
    end
    
    # wait for all tasks to complete
    for task in producer_tasks
        wait(task)
    end
    for task in consumer_tasks
        wait(task)
    end
    
    close(task_channel)
    
    end_time = time_ns()
    _ = processed[]  # prevent optimization
    return (end_time - start_time) / 1_000_000
end

# fibonacci computation
function fibonacci(n::Int)::Int64
    if n <= 1
        return Int64(n)
    end
    
    a, b = Int64(0), Int64(1)
    for i in 2:n
        a, b = b, a + b
    end
    return b
end

# parallel mathematical work test using Julia's @threads
function parallel_math_test(num_threads::Int, work_per_thread::Int)::Float64
    start_time = time_ns()
    
    total_sum = Threads.Atomic{Int64}(0)
    
    @sync begin
        for i in 1:num_threads
            @async begin
                local_sum = Int64(0)
                
                for j in 1:work_per_thread
                    local_sum += fibonacci(35)
                    
                    # additional mathematical work
                    for k in 1:1000
                        local_sum += Int64(k * k)
                    end
                end
                
                Threads.atomic_add!(total_sum, local_sum)
            end
        end
    end
    
    end_time = time_ns()
    _ = total_sum[]  # prevent optimization
    return (end_time - start_time) / 1_000_000
end

# async file processing test
function async_file_test(num_files::Int)::Float64
    start_time = time_ns()
    
    temp_dir = mktempdir(prefix="concurrency_test_")
    processed = Threads.Atomic{Int}(0)
    
    @sync begin
        for i in 1:num_files
            @async begin
                try
                    filename = joinpath(temp_dir, "test_$(i).dat")
                    
                    # write file
                    open(filename, "w") do f
                        for j in 1:1000
                            println(f, "data_$(i)_$(j)")
                        end
                    end
                    
                    # read and process file
                    lines = readlines(filename)
                    
                    # simulate processing
                    line_count = length(lines)
                    if line_count > 0
                        # process each line
                        for line in lines
                            _ = length(line)  # simulate processing
                        end
                        Threads.atomic_add!(processed, 1)
                    end
                    
                    # cleanup
                    rm(filename, force=true)
                    
                catch e
                    # ignore errors for benchmark
                end
            end
        end
    end
    
    # cleanup temp directory
    rm(temp_dir, recursive=true, force=true)
    
    end_time = time_ns()
    _ = processed[]  # prevent optimization
    return (end_time - start_time) / 1_000_000
end

# thread pool worker task
function pool_worker_task()::Int
    # simulate varied workload
    work = 0
    for j in 1:10000
        work += j * j
    end
    
    sleep(0.0001)  # 100 microseconds
    return 1
end

# thread pool performance test using Julia's @distributed
function thread_pool_test(pool_size::Int, total_tasks::Int)::Float64
    start_time = time_ns()
    
    completed = Threads.Atomic{Int}(0)
    
    # use Julia's task scheduling with limited concurrency
    @sync begin
        semaphore = Base.Semaphore(pool_size)
        
        for i in 1:total_tasks
            @async begin
                Base.acquire(semaphore)
                try
                    result = pool_worker_task()
                    Threads.atomic_add!(completed, result)
                finally
                    Base.release(semaphore)
                end
            end
        end
    end
    
    end_time = time_ns()
    _ = completed[]  # prevent optimization
    return (end_time - start_time) / 1_000_000
end

function main()
    scale_factor = 1
    
    if length(ARGS) > 0
        try
            scale_factor = parse(Int, ARGS[1])
            if scale_factor <= 0
                scale_factor = 1
            end
        catch
            println(stderr, "Invalid scale factor. Using default 1.")
            scale_factor = 1
        end
    end
    
    total_time = 0.0
    
    # warm up Julia JIT compiler
    parallel_http_test(2)
    producer_consumer_test(1, 10)
    parallel_math_test(1, 5)
    async_file_test(2)
    thread_pool_test(2, 10)
    
    # actual benchmarks
    total_time += parallel_http_test(50 * scale_factor)
    total_time += producer_consumer_test(4, 1000 * scale_factor)
    total_time += parallel_math_test(4, 100 * scale_factor)
    total_time += async_file_test(20 * scale_factor)
    total_time += thread_pool_test(8, 500 * scale_factor)
    
    @printf "%.3f\n" total_time
end

main()

File: speed/concurrency/concurrency.nim
import std/[asyncdispatch, httpclient, times, os, strutils, threadpool, locks, channels]
import std/[strformat, tempfiles, random]

# parallel http requests test using async/await
proc parallelHttpTest(numRequests: int): float =
    let start = epochTime()
    var successful = 0
    
    proc makeRequest(): Future[bool] {.async.} =
        try:
            let client = newAsyncHttpClient()
            defer: client.close()
            
            let response = await client.get("http://127.0.0.1:8000/fast")
            if response.status.startsWith("200"):
                return true
        except:
            discard
        return false
    
    var futures: seq[Future[bool]]
    for i in 0 ..< numRequests:
        futures.add(makeRequest())
    
    for future in futures:
        if waitFor(future):
            inc successful
    
    let duration = epochTime() - start
    discard successful # prevent optimization
    return duration * 1000.0

# producer-consumer queue test using channels
proc producerConsumerTest(numPairs: int, itemsPerThread: int): float =
    let start = epochTime()
    var 
        taskChannel: Channel[int]
        processed = 0
        processedLock: Lock
    
    taskChannel.open()
    initLock(processedLock)
    
    proc producer(producerId: int) {.thread.} =
        for i in 0 ..< itemsPerThread:
            taskChannel.send(producerId * 1000 + i)
    
    proc consumer() {.thread.} =
        var localCount = 0
        for i in 0 ..< itemsPerThread:
            let item = taskChannel.recv()
            
            # simulate processing
            let dummy = item * item
            inc localCount
        
        withLock processedLock:
            processed += localCount
    
    var threads: seq[Thread[int]]
    newSeq(threads, numPairs * 2)
    
    # create producer threads
    for i in 0 ..< numPairs:
        createThread(threads[i], producer, i)
    
    # create consumer threads
    for i in 0 ..< numPairs:
        createThread(threads[numPairs + i], consumer)
    
    # wait for all threads to complete
    for thread in threads:
        joinThread(thread)
    
    taskChannel.close()
    deinitLock(processedLock)
    
    let duration = epochTime() - start
    discard processed # prevent optimization
    return duration * 1000.0

# fibonacci computation
proc fibonacci(n: int): int64 =
    if n <= 1:
        return n.int64
    
    var a, b = (0i64, 1i64)
    for i in 2..n:
        let temp = a + b
        a = b
        b = temp
    
    return b

# mathematical worker function
proc mathWorker(workPerThread: int): int64 =
    var localSum = 0i64
    
    for i in 0 ..< workPerThread:
        localSum += fibonacci(35)
        
        # additional mathematical work
        for k in 0 ..< 1000:
            localSum += (k * k).int64
    
    return localSum

# parallel mathematical work test using spawn
proc parallelMathTest(numThreads: int, workPerThread: int): float =
    let start = epochTime()
    
    var futures: seq[FlowVar[int64]]
    for i in 0 ..< numThreads:
        futures.add(spawn mathWorker(workPerThread))
    
    var totalSum = 0i64
    for future in futures:
        totalSum += ^future
    
    let duration = epochTime() - start
    discard totalSum # prevent optimization
    return duration * 1000.0

# async file processing test
proc processFileAsync(fileId: int, tempDir: string): Future[bool] {.async.} =
    try:
        let filename = tempDir / fmt"test_{fileId}.dat"
        
        # write file
        var content = ""
        for j in 0 ..< 1000:
            content.add(fmt"data_{fileId}_{j}" & "\n")
        
        writeFile(filename, content)
        
        # read and process file
        let fileContent = readFile(filename)
        let lineCount = fileContent.count('\n')
        
        if lineCount > 0:
            # cleanup
            removeFile(filename)
            return true
        
    except:
        discard
    
    return false

proc asyncFileTest(numFiles: int): float =
    let start = epochTime()
    let tempDir = createTempDir("concurrency_test_", "")
    
    var futures: seq[Future[bool]]
    for i in 0 ..< numFiles:
        futures.add(processFileAsync(i, tempDir))
    
    var processed = 0
    for future in futures:
        if waitFor(future):
            inc processed
    
    # cleanup temp directory
    removeDir(tempDir)
    
    let duration = epochTime() - start
    discard processed # prevent optimization
    return duration * 1000.0

# thread pool worker task
proc poolWorkerTask(): int =
    # simulate varied workload
    var work = 0
    for j in 0 ..< 10000:
        work += j * j
    
    sleep(1) # 1 millisecond (Nim's sleep is in milliseconds)
    return 1

# thread pool performance test using threadpool
proc threadPoolTest(poolSize: int, totalTasks: int): float =
    let start = epochTime()
    setMaxPoolSize(poolSize)
    
    var futures: seq[FlowVar[int]]
    for i in 0 ..< totalTasks:
        futures.add(spawn poolWorkerTask())
    
    var completed = 0
    for future in futures:
        completed += ^future
    
    let duration = epochTime() - start
    discard completed # prevent optimization
    return duration * 1000.0

proc main() =
    var scaleFactor = 1
    
    if paramCount() > 0:
        try:
            scaleFactor = parseInt(paramStr(1))
            if scaleFactor <= 0:
                scaleFactor = 1
        except:
            echo "Invalid scale factor. Using default 1."
            scaleFactor = 1
    
    var totalTime = 0.0
    
    totalTime += waitFor(parallelHttpTest(50 * scaleFactor))
    totalTime += producerConsumerTest(4, 1000 * scaleFactor)
    totalTime += parallelMathTest(4, 100 * scaleFactor)
    totalTime += waitFor(asyncFileTest(20 * scaleFactor))
    totalTime += threadPoolTest(8, 500 * scaleFactor)
    
    echo fmt"{totalTime:.3f}"

when isMainModule:
    main()